lib/primer/view_components/linters/migrations/button_component.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Migrations
+      # Lint and hopefully autocorrect cases of the deprecated ButtonComponent
+      class ButtonComponent < RuboCop::Cop::Cop
+        INVALID_MESSAGE = <<~STR
+          `Primer::ButtonComponent` is deprecated. Please use `Primer::Beta::Button` instead.
+        STR
+
+        def_node_matcher :button_component, <<~PATTERN
+          (send $(const (const nil? :Primer) :ButtonComponent) :new ...)
+        PATTERN
+
+        def_node_matcher :hash_with_dropdown_value?, <<~PATTERN
+          (hash ... (pair (sym :dropdown) (...)) ... )
+        PATTERN
+
+        def_node_matcher :hash_with_group_item_value?, <<~PATTERN
+          (hash ... (pair (sym :group_item) (...)) ... )
+        PATTERN
+
+        def_node_matcher :hash_with_variant_value, <<~PATTERN
+          (hash ... (pair $(sym :variant) (...)) ... )
+        PATTERN
+
+        def_node_matcher :hash_with_scheme_outline_value, <<~PATTERN
+          (hash ... (pair (sym :scheme) $(...)) ... )
+        PATTERN
+
+        def on_send(node)
+          return unless button_component(node)
+
+          add_offense(node, message: INVALID_MESSAGE)
+        end
+
+        def autocorrect(node)
+          return if hash_with_dropdown_value?(node.arguments.first)
+          return if hash_with_group_item_value?(node.arguments.first)
+
+          lambda do |corrector|
+            corrector.replace(button_component(node), "Primer::Beta::Button")
```

Comment:
Sometimes it's this easy, but also sometimes it isn't. For example,

```erb
<%= render Primer::ButtonComponent.new() do %>
    <%= primer_octicon(:alert) %>
    Alert button
<% end %>
```

Would need to become

```erb
<%= render Primer::Beta::Button.new() do |b| %>
    <% b.with_leading_visual_icon(:alert) %>
    Alert Button
<% end %>
```

That seems like a lot to ask the linter to do. Maybe we fail out early if it has more than just text inside? or maybe I'm thinking about it too hard and we just go with it

Reply:
🤔 I guess it wouldn't be terrible

**Beta::Button the correct way**
<img width="115" alt="image" src="https://github.com/primer/view_components/assets/54012/fdecc112-6fd7-4e2d-8794-438fc8ac8bd1">

**Beta::Button with improper icon**
<img width="112" alt="image" src="https://github.com/primer/view_components/assets/54012/9e80264a-7c4a-4937-b68f-be0c87b0c200">

@langermank could you weigh in on if this is 🚫 ?

Reply:
Yeah, this is a blocker for sure. Sorry ya'll 😛 icons need to be in their slots (leading/trailing/trailing action)

Reply:
Do we have any of these in GH? I took a look and can only find buttons that use `primer_octocon` that should be iconbuttons instead (perhaps this is another lint we can add).

Reply:
Probably a better example might be this ["Edit Pins" button with the trailing action](https://github.com/github/github/blob/2d2104e1bce33c5f0950536e814656144750bae2/app/components/repositories/pin_organization_repo_component.html.erb#L6-L10) which would need to change like this 

```diff
- <%= render Primer::ButtonComponent.new(tag: :summary, variant: :small) do |c| %>
+ <%= render Primer::Beta::Button.new(tag: :summary, size: :small) do |c| %>
    <% c.with_leading_visual_icon(icon: :pin) %>
    <%= "Edit Pins" %>
-    <span class="dropdown-caret"></span>
+    <% c.with_trailing_action_icon(icon: "triangle-down") %>
<% end %>
```